Add SignedData func

smallhive opened the pull request:
closes #510

Added func to known places with `Sign` method, plus for some places when I think it would be useful.
Of course, in some way it is almost similar to `Marshal` func, but for consistency called it _SignedData_.

If I missed any place, just mention it

roman-khimov commented on container/container.go:
```diff
@@ -155,6 +155,11 @@ func (x Container) Marshal() []byte {
 	return x.v2.StableMarshal(nil)
 }
 
+// SignedData returns actual payload which would be signed, if you call [client.Client.ContainerPut] method.
```
`... call [Container.CalculateSignature] ...`?

smallhive replied:
Updated

cthulhu-rider reviewed (commented):
what about `object.Object`?

---

also suggest to add unified test to `crypto/test`:
```go
type SignedComponent interface {
	SignedData() []byte
	Sign(neofscrypto.Signer) error
	VerifySignature() bool
}

func TestSignedData(tb testing.TB, signer neofscrypto.Signer, cmp SignedComponent) {
	data := cmp.SignedData()
	
	sig, err := signer.Sign(data)
	require.NoError(tb, err)
	
	err = cmp.Sign(neofscrypto.NewStaticSigner(signer.Scheme(), sig, signer.Public()))
	require.NoError(tb, err)
	
	require.True(tb, cmp.VerifySignature())
}
```
and test this against all touched components

carpawell reviewed (commented):
`SignedData returns data that _would be signed_` breaks me a little. Mabe `SignedData returns the data that is signed after a [Token.Sign] call`?

smallhive commented on bearer/bearer.go:
```diff
@@ -268,6 +268,11 @@ func (b *Token) Sign(signer neofscrypto.Signer) error {
 	return nil
 }
 
+// SignedData returns actual payload which would be signed, if you call [Token.Sign] method.
```
Updated

smallhive commented:
> what about `object.Object`?

Nothing, it has a different API for signing with different names. I tend to think we may rewrite it to be compatible with `SignedComponent` interface. I did these changes in the last commit

I've added the suggested test, but we had some changes according to this

cthulhu-rider commented:
> `SignedData returns data that _would be signed_` breaks me a little. Mabe `SignedData returns the data that is signed after a [Token.Sign] call`?

```
// SignedData returns signed data of the T.
//
// See also [T.Sign].
```
?

cthulhu-rider commented on crypto/test/tests.go:
```diff
@@ -32,3 +32,25 @@ func RandomSignerRFC6979(tb testing.TB) user.Signer {
+}
+
+// SignedDataComponent tests [SignedComponent] for valid data generation by SignedData function.
+func SignedDataComponent(tb testing.TB, signer user.Signer, cmp SignedComponent) {
```
`signer` may be of type `neofscrypto.Signer`, isn't it?

smallhive replied:
No, it can't be. If it `neofscrypto.Signer` we cover just couple of structs

cthulhu-rider replied:
it's not about structs, it's about function body. It deals with `user.Signer` as `neofscrypto.Signer`

cthulhu-rider commented on bearer/bearer.go:
```diff
@@ -253,8 +253,8 @@ func (b Token) AssertUser(id user.ID) bool {
-// See also VerifySignature, Issuer.
-func (b *Token) Sign(signer neofscrypto.Signer) error {
+// See also VerifySignature, Issuer, [Token.SignedData].
+func (b *Token) Sign(signer user.Signer) error {
```
dont do this just for test, make a workaround for the test if needed

smallhive replied:
In this case we should duplicate `SignedComponent` for `neofscrypto.Signer` and for `user.Signer`

roman-khimov replied:
But why? This is certainly not worth a change for test. And tests should be fine with `neofscrypto.Signer` since `user.Signer` includes it.

roman-khimov commented on bearer/bearer.go:
```diff
@@ -253,7 +253,7 @@ func (b Token) AssertUser(id user.ID) bool {
 // expected to be calculated as a final stage of Token formation.
 //
-// See also VerifySignature, Issuer.
+// See also VerifySignature, Issuer, [Token.SignedData].
```
Doc links should be fixed at the same time.

smallhive replied:
updated

roman-khimov commented on bearer/bearer.go:
```diff
@@ -268,6 +268,13 @@ func (b *Token) Sign(signer neofscrypto.Signer) error {
 	return nil
 }
 
+// SignedData returns actual payload for sign.
```
`to sign`? I always remember @EliChin at these moments.

smallhive replied:
updated

roman-khimov commented on container/container.go:
```diff
@@ -467,7 +474,7 @@ func (x Container) ReadDomain() (res Domain) {
 // can be used.
 //
-// See also VerifySignature.
+// See also VerifySignature, [Container.SignedData].
```
Link (and elsewhere).

smallhive replied:
Updated

roman-khimov commented on crypto/test/tests.go:
```diff
@@ -32,3 +32,48 @@ func RandomSignerRFC6979(tb testing.TB) user.Signer {
+// SignedComponent describes component which can signed and the signature may be verified.
+type SignedComponent interface {
+	SignedData() []byte
+	Sign(neofscrypto.Signer) error
```
Likely some generic magic can be used here, but two interfaces are kinda OK for test as well.

roman-khimov commented on object/fmt.go:
```diff
@@ -93,9 +93,10 @@ func (o *Object) VerifyID() error {
-// CalculateAndSetSignature signs id with provided key and sets that signature to
-// the object.
-func (o *Object) CalculateAndSetSignature(signer neofscrypto.Signer) error {
+// Sign signs id with provided key and sets that signature to the object.
```
`signs object ID`?

smallhive replied:
updated

smallhive commented:
Updated. Make two test interfaces/functions for each `user.Signer` and `neofscrypto.Signer`. But Object functions I renamed to be compatible with new test interface and more important with other signable components

smallhive commented:
No problem with the signer it is OK. Problem with checking functions
```
func SignedDataComponent(tb testing.TB, signer neofscrypto.Signer, cmp SignedComponent)

func SignedDataComponentUser(tb testing.TB, signer user.Signer, cmp SignedComponentUserSigner)
```

They don't want to accept the wrong types for `cmp` parameter.

```
Cannot use 'val' (type *session.Container) as the type SignedComponent Type does not implement 'SignedComponent'
need the method: Sign(neofscrypto.Signer) error 
have the method: Sign(signer user.Signer) error
```

smallhive commented:
That is why we have two interfaces and two functions to check them

roman-khimov commented on bearer/bearer.go:
```diff
@@ -244,16 +244,16 @@ func (b Token) AssertUser(id user.ID) bool {
 // NeoFS API V2 protocol.
 //
-// Note that any Token mutation is likely to break the signature, so it is
+// Note that any [Token] mutation is likely to break the signature, so it is
```
These are not related, of course. I meant only those lines that were changed. Can we separate these changes into a different commit?

smallhive replied:
We can, but why not related? Updating documentation in the same place as `See also [Token.VerifySignature], [Token.Issuer], [Token.SignedData].` What the difference between this row and other documentation in this function?

roman-khimov replied:
You're inevitably touching "See also [Token.VerifySignature], [Token.Issuer], [Token.SignedData]" in your PR, so you can change links as well. You don't need to touch these comments to deal with `SignedData`, that's why it's a separate change. Good one! But separate.

smallhive replied:
OK, I see, I will separate these changes

smallhive replied:
Moved to separate commit

cthulhu-rider commented:
```
need the method: Sign(neofscrypto.Signer) error 
have the method: Sign(signer user.Signer) error
```
needed method is more generic and may be implemented in-place

cthulhu-rider commented on crypto/test/tests.go:
```diff
@@ -32,3 +32,48 @@ func RandomSignerRFC6979(tb testing.TB) user.Signer {
+
+	static := neofscrypto.NewStaticSigner(signer.Scheme(), sig, signer.Public())
+
+	err = cmp.Sign(user.NewSigner(static, signer.UserID()))
```
does `VerifySignature` check passed user? If not, then `SignedDataComponent` would be enough (we may provide fast cast from `SignedComponentUserSigner` to `SignedComponent`

roman-khimov replied:
I see what you mean with wrapping, but it's a test code and it can be improved later.
